plugins/inputs/t128_tank/reader.go

```diff
@@ -138,24 +134,41 @@ func (r *Reader) Run(mainCtx context.Context) {
 		readCtxCancel()
 		return
 	}
-
 	nextSaveCheck := time.NewTicker(2 * time.Second)
-	defer readCtxCancel()
+	defer func() {
+		readCtxCancel()
+		close(r.lastObservedIndex)
+		observedValue, ok := <-r.lastObservedIndex
+		if ok {
+			r.setIndex(r.indexPath, index{value: observedValue})
+		} else {
+			r.setIndex(r.indexPath, index{value: lastIndex.value})
+		}
+	}()
 	go r.read(readCtx, lastIndex.next())
 
 	for {
 
 		select {
 		case <-mainCtx.Done():
 			r.log.Errorf("%s reader done", r.topic)
-			r.saveIndex(r.lastSavedIndex)
 			return
+		case observedValue := <-r.lastObservedIndex:
+			lastIndex.value = observedValue
+			if lastIndex.value%1000 == 0 {
+				r.setIndex(r.indexPath, index{value: lastIndex.value})
+			}
 		case <-nextSaveCheck.C:
-			if lastIndex.value > r.lastSavedIndex {
-				r.saveIndex(lastIndex.value)
+			var observedValue uint64
+			select {
+			case observedValue = <-r.lastObservedIndex:
```

Comment:
1. This isn't correct because the whole point of this ticker is to check when observed values aren't coming. In that case, the `default -> continue` path will be taken and the index won't be saved.
2. Don't bother with the observed index at all in this case block. It has its own case.